avr-hal-generic/src/i2c.rs

```diff
@@ -329,51 +336,110 @@ macro_rules! impl_twi_i2c {
-                //    that clearing this flag starts the operation of the TWI, so all accesses to the TWI Address Register (TWAR), TWI
-                //    Status Register (TWSR), and TWI Data Register (TWDR) must be complete before clearing this flag.
+                //    This bit is set by hardware when the TWI has finished its current job and
+                //    expects application software response.  If the I-bit in SREG and TWIE in TWCR
+                //    are set, the MCU will jump to the TWI Interrupt Vector. While the TWINT Flag
+                //    is set, the SCL low period is stretched. The TWINT Flag must be cleared by
+                //    software by writing a logic one to it. Note that this flag is not
+                //    automatically cleared by hardware when executing the interrupt routine. Also
+                //    note that clearing this flag starts the operation of the TWI, so all accesses
+                //    to the TWI Address Register (TWAR), TWI Status Register (TWSR), and TWI Data
+                //    Register (TWDR) must be complete before clearing this flag.
                 //
                 //
                 //    Bit 5 – TWSTA: TWI START Condition Bit
                 //
-                //    The application writes the TWSTA bit to one when it desires to become a Master on the 2-wire Serial Bus. The
-                //    TWI hardware checks if the bus is available, and generates a START condition on the bus if it is free. However,
-                //    if the bus is not free, the TWI waits until a STOP condition is detected, and then generates a new START
-                //    condition to claim the bus Master status. TWSTA must be cleared by software when the START condition has
-                //    been transmitted.
+                //    The application writes the TWSTA bit to one when it desires to become a
+                //    Master on the 2-wire Serial Bus. The TWI hardware checks if the bus is
+                //    available, and generates a START condition on the bus if it is free. However,
+                //    if the bus is not free, the TWI waits until a STOP condition is detected, and
+                //    then generates a new START condition to claim the bus Master status. TWSTA
+                //    must be cleared by software when the START condition has been transmitted.
                 //
                 //    NOTE: `twstart` macro token is defined as `twsta` in chips/[CHIP]/src/lib.rs
                 self.p.$twcr.write(|w| w
                     .$twen().set_bit()
                     .$twint().set_bit()
                     .$twstart().set_bit()
                 );
-                // NOTE: why do we need to wait here? `wait()` is just reading twcr.twint to see if
-                // the bit is set...but it was just set in the line above...
+
+                // QUESTION: why do we need to wait here? `wait()` is just reading twcr.twint to see if
```

Comment:
Why do we need the wait() here?

Reply:
Register bits don't necessarily act like memory cells.  The following things which you might assume from a memory cell do _not_ always apply for register bits:

- Reading a register twice without any writes in between will yield the same value.

  _Counter Examples_:
  - A FIFO register will return the next byte from the internal buffer on each read.  I.e. the first read will yield the first byte, the second read will yield the second byte, etc.
  - Sometimes reading a bit will reset it back to zero.

- Writing a value to a register and reading it back will give back the same value.

  _Counter Examples_: In hardware, the logic that receives data written to a register is entirely disjunct from the logic presenting data when the register is read.  There are many different ways in which this shows:

  - Sometimes only a subset of bits in the register are writeable (=read-only).  All other bits from the written value will be ignored.
  - Some bits are write-only and reading will always return 0.
  - Sometimes writing a 1 to a bit will reset it back to zero.
  - Sometimes a bits value is entirely dependent on hardware state and writing it has a completely different effect than what reading is for (bits are "reused" for two different purposes).

---

In this specific case, we write `twint` to 1 which will clear the flag:

> The TWINT Flag must be cleared by software by writing a logic one to it.

Then we `wait()` until hardware sets it again.

Reply:
Great comment. Do you mind if I capture this in a Newbie FAQ?

Reply:
Well, this is true for the entirety of embedded software development so I don't think `avr-hal` is the right place for it.  But if you want to post it somewhere else, please go a head :)

Reply:
Gotcha. I'll probably post a summary of this experience on my blog. No one reads it, but perhaps it'll help someone someday...probably me in 6 months :-)